### PR TITLE
feat(core): Rate limit forgot password endpoint

### PR DESCRIPTION
**Pull request by RicardoE105**
Github issue / Community forum post (link here to close automatically):

### CODE REVIEW

**Review by netroy — COMMENTED** [review]
Should we also update the frontend to handle `429`s and show some other message instead of 
> Please contact your administrator (problem with your SMTP setup)

**Comment by csuermann** [comment]
Great PR! Just a few comments:
- What will the user experience look like when the rate-limit kicks in?
- We should consider deployment scenarios with more than 1 primary n8n instance and consider externalizing the rate-limiter's state to Redis in those cases
- Will rate-limiting based on IP addresses also work reliably if requests come in through proxy servers or load balancers? ([Docs](https://express-rate-limit.mintlify.app/guides/troubleshooting-proxy-issues))

**Comment by netroy** [comment]
> Great PR! Just a few comments:
> 
>     * What will the user experience look like when the rate-limit kicks in?
> 
>     * We should consider deployment scenarios with more than 1 primary n8n instance and consider externalizing the rate-limiter's state to Redis in those cases
> 
>     * Will rate-limiting based on IP addresses also work reliably if requests come in through proxy servers or load balancers? ([Docs](https://express-rate-limit.mintlify.app/guides/troubleshooting-proxy-issues))

**Comment by netroy** [comment]
Sorry, clicked on the wrong button.

**Comment by RicardoE105** [comment]
> Great PR! Just a few comments:
> 
> * What will the user experience look like when the rate-limit kicks in?
> * We should consider deployment scenarios with more than 1 primary n8n instance and consider externalizing the rate-limiter's state to Redis in those cases
> * Will rate-limiting based on IP addresses also work reliably if requests come in through proxy servers or load balancers? ([Docs](https://express-rate-limit.mintlify.app/guides/troubleshooting-proxy-issues))

1. as it's right now, it would show "Please contact your administrator (problem with your SMTP setup)", but Adi suggested above to show a proper error.
2. Had a small chat with @netroy about this and this limiting would still work just that each instance would its individual count. @netroy any other thought here?
3. Ah very good question. Had not thought about this. The solution they present have you find out the numbers of trusted proxy manually?! Will have a sync with the cloud team to see how we can address this at least taking in consideration our infra

File: packages/cli/src/controllers/passwordReset.controller.ts [hunk 1]
```diff
@@ -46,7 +47,14 @@ export class PasswordResetController {
 	/**
 	 * Send a password reset email.
 	 */
-	@Post('/forgot-password')
+	@Post('/forgot-password', {
+		middlewares: [
+			rateLimit({
```
Comment (netroy) [inline packages/cli/src/controllers/passwordReset.controller.ts:52]: this middleware should be disabled during tests.

File: packages/cli/src/controllers/passwordReset.controller.ts [hunk 2]
```diff
@@ -46,7 +47,14 @@ export class PasswordResetController {
...
+				windowMs: 15 * 60 * 1000, // 15 minutes
+				limit: 10, // Limit each IP to 10 requests per `window` (here, per 15 minutes).
```
Comment (netroy) [inline packages/cli/src/controllers/passwordReset.controller.ts:54]: Can we change these to 
```suggestion
				windowMs: 5 * 60 * 1000, // 5 minutes
				limit: 5, // Limit each IP to 5 requests per `window` (here, per 5 minutes).
```

**Review by netroy — CHANGES_REQUESTED** [review]

File: packages/cli/src/decorators/Route.ts [hunk 3]
```diff
@@ -13,10 +17,13 @@ const RouteFactory =
 		const controllerClass = target.constructor;
 		const routes = (Reflect.getMetadata(CONTROLLER_ROUTES, controllerClass) ??
 			[]) as RouteMetadata[];
+		const middlewares = options.middlewares ?? [];
+		if (options.throttle && !inTest) middlewares.unshift(rateLimit(options.throttle));
```
Comment (netroy) [inline packages/cli/src/decorators/Route.ts:21]: do we have any other endpoints where we'd like to add throttling as well? if not, I think we should keep this explicitly in `PasswordResetController` for now, until we a need for this elsewhere.

Comment (RicardoE105) [reply packages/cli/src/decorators/Route.ts:21]: Wanted to avoid that because it looks very ugly, but makes sense! Will reverse the last commit

**Review by netroy — COMMENTED** [review]
LGTM. 
But, should we also add the new config variable in schema
```
	proxy_hops: {
		format: Number,
		default: 0,
		env: 'N8N_PROXY_HOPS',
		doc: 'Number of reverse-proxies n8n is running behind',
	},
```

and update the code after [this line](https://github.com/n8n-io/n8n/blob/master/packages/cli/src/AbstractServer.ts#L56) to include
```typescript
const proxyHops = config.getEnv('proxy_hops');
if (proxyHops > 0) this.app.set('trust proxy', proxyHops);
```

**Review by netroy — APPROVED** [review]
I suppose we should make `windowMs` and `limit` configurable. But, I also think that these defaults should be enough for almost everyone, and we could consider making these configurable if/when there is an actual demand for it.

Thanks for patching this so quickly 🙏🏽

**Comment by RicardoE105** [comment]
Thanks for all the help @netroy 🙏

**Comment by janober** [comment]
Got released with [`n8n@1.16.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.16.0)

**Pull request merged**